pkg/hubble/relay/observer.go

```diff
@@ -20,17 +20,209 @@ import (
 	"time"
 
 	observerpb "github.com/cilium/cilium/api/v1/observer"
+	relaypb "github.com/cilium/cilium/api/v1/relay"
 	"github.com/cilium/cilium/pkg/hubble/relay/queue"
-	"github.com/sirupsen/logrus"
+	nodeTypes "github.com/cilium/cilium/pkg/node/types"
 
+	"github.com/golang/protobuf/ptypes"
+	"github.com/sirupsen/logrus"
 	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
 )
 
 // ensure that Server implements the observer.ObserverServer interface.
 var _ observerpb.ObserverServer = (*Server)(nil)
 
+func retrieveFlowsFromPeer(
+	ctx context.Context,
+	conn *grpc.ClientConn,
+	req *observerpb.GetFlowsRequest,
+	flows chan<- *observerpb.GetFlowsResponse,
+) error {
+	client := observerpb.NewObserverClient(conn)
+	c, err := client.GetFlows(ctx, req)
+	if err != nil {
+		return err
+	}
+	for {
+		flow, err := c.Recv()
+		switch err {
+		case io.EOF, context.Canceled:
+			return nil
+		case nil:
+			select {
+			case flows <- flow:
+			case <-ctx.Done():
+				return nil
+			}
+		default:
+			if status.Code(err) != codes.Canceled {
+				return err
+			}
+			return nil
+		}
+	}
+}
+
+func sortFlows(
+	ctx context.Context,
+	flows <-chan *observerpb.GetFlowsResponse,
+	qlen int,
+	bufferDrainTimeout time.Duration,
+) <-chan *observerpb.GetFlowsResponse {
+	pq := queue.NewPriorityQueue(qlen)
+	sortedFlows := make(chan *observerpb.GetFlowsResponse, qlen)
+
+	go func() {
+		defer close(sortedFlows)
+
+	flowsLoop:
+		for {
+			select {
+			case flow, ok := <-flows:
+				if !ok {
+					break flowsLoop
+				}
+				if pq.Len() == qlen {
+					f := pq.Pop()
+					select {
+					case sortedFlows <- f:
+					case <-ctx.Done():
+						return
+					}
+				}
+				pq.Push(flow)
+			case <-time.After(bufferDrainTimeout): // make sure to drain the queue when no new flow responses are received
+				if f := pq.Pop(); f != nil {
+					select {
+					case sortedFlows <- f:
+					case <-ctx.Done():
+						return
+					}
+				}
+			case <-ctx.Done():
+				return
+			}
+		}
+		// drain the queue
+		for f := pq.Pop(); f != nil; f = pq.Pop() {
+			select {
+			case sortedFlows <- f:
+			case <-ctx.Done():
+				return
+			}
+		}
+
+	}()
+
+	return sortedFlows
+}
+
+func nodeStatusError(err error, nodeNames ...string) *observerpb.GetFlowsResponse {
+	msg := err.Error()
+	if s, ok := status.FromError(err); ok && s.Code() == codes.Unknown {
```

Comment:
Why only if `s.Code() == codes.Unknown`?

Reply:
It's mirroring the logic we have on the client-side: 

https://github.com/cilium/hubble/blob/9bca65beb242f467d869f044f54b58cf26e14c88/cmd/observe/observe.go#L90-L94

Unknown gRPC codes are errors thrown by the Hubble Observer for application-level errors such as `invalid IP address in filter`. Stripping out the `grpc error: code = Unknown desc =` makes this easier to understand for users, showing them that this is a problem with the request, not with grpc. 

All other codes (e.g. Unimplemented) tend to be protocol-level errors, therefore it makes sense to preserve the information that this is coming from the gRPC stack.